### PR TITLE
LibJS: Implement the rest of the ArrayBuffer transfer proposal

### PR DESCRIPTION
**Pull request by PrestonLTaylor**
[Proposal Link](https://tc39.es/proposal-arraybuffer-transfer/)

### CODE REVIEW

File: Userland/Libraries/LibJS/Print.cpp [hunk 1]
```diff
@@ -384,9 +384,15 @@ ErrorOr<void> print_promise(JS::PrintContext& print_context, JS::Promise const&
 
 ErrorOr<void> print_array_buffer(JS::PrintContext& print_context, JS::ArrayBuffer const& array_buffer, HashTable<JS::Object*>& seen_objects)
 {
+    TRY(print_type(print_context, "ArrayBuffer"sv));
+
+    if (array_buffer.is_detached()) {
```
Comment (linusg) [inline Userland/Libraries/LibJS/Print.cpp:389]: Detaching an array buffer sets its byte length to 0 and accessing it afterwards should not crash - so I'd argue the correct fix is handling detached array buffers in the `byte_length()` getter.

Comment (linusg) [reply Userland/Libraries/LibJS/Print.cpp:389]: (we can still print out the detached state if you want to repurpose this commit)

File: Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp [hunk 2]
```diff
@@ -128,6 +130,28 @@ JS_DEFINE_NATIVE_FUNCTION(ArrayBufferPrototype::slice)
     return new_array_buffer_object;
 }
 
+// 25.1.5.5 ArrayBuffer.prototype.transfer ( [ newLength ] ), https://tc39.es/proposal-arraybuffer-transfer/#sec-arraybuffer.prototype.transfer
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp:133]: Please keep these in spec order - it was already a bit messy before, but now we have 3, 5, 6, 1, 4. Let's put these two after 4 and we can fix up the rest later.

File: Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp [hunk 3]
```diff
@@ -128,6 +130,28 @@ JS_DEFINE_NATIVE_FUNCTION(ArrayBufferPrototype::slice)
...
+JS_DEFINE_NATIVE_FUNCTION(ArrayBufferPrototype::transfer)
+{
+    // 1. Let O be the this value.
+    auto array_buffer_object = TRY(typed_this_value(vm));
+
+    // 2. Return ? ArrayBufferCopyAndDetach(O, newLength, preserve-resizability).
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp:139]:
```suggestion
    // 2. Return ? ArrayBufferCopyAndDetach(O, newLength, preserve-resizability).
```

(not sure what OS/browser you're using, but we commonly replace the NBSPs from the spec with regular spaces, some do that automatically when copying)

File: Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp [hunk 4]
```diff
@@ -128,6 +130,28 @@ JS_DEFINE_NATIVE_FUNCTION(ArrayBufferPrototype::slice)
...
+    return TRY(copy_and_detach(vm, array_buffer_object, new_length, PreserveResizability::Preserve));
+}
+
+// 25.1.5.6 ArrayBuffer.prototype.transferToFixedLength ( [ newLength ] ), https://tc39.es/proposal-arraybuffer-transfer/#sec-arraybuffer.prototype.transfertofixedlength
+JS_DEFINE_NATIVE_FUNCTION(ArrayBufferPrototype::transfer_to_fixed_length)
+{
+    // 1. Let O be the this value.
+    auto array_buffer_object = TRY(typed_this_value(vm));
+
+    // 2. Return ? ArrayBufferCopyAndDetach(O, newLength, preserve-resizability).
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp:150]:
```suggestion
    // 2. Return ? ArrayBufferCopyAndDetach(O, newLength, fixed-length).
```

File: Userland/Libraries/LibJS/Runtime/ArrayBuffer.h [hunk 5]
```diff
@@ -22,6 +22,11 @@ struct ClampedU8 {
 // 25.1.1 Notation (read-modify-write modification function), https://tc39.es/ecma262/#sec-arraybuffer-notation
 using ReadWriteModifyFunction = Function<ByteBuffer(ByteBuffer, ByteBuffer)>;
 
+enum class PreserveResizability {
+    Fixed,
+    Preserve
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayBuffer.h:27]: Always match spec names as closely as possible

```suggestion
    FixedLength,
    PreserveResizability
```

File: Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp [hunk 6]
```diff
@@ -182,4 +182,52 @@ ThrowCompletionOr<ArrayBuffer*> clone_array_buffer(VM& vm, ArrayBuffer& source_b
     return target_buffer;
 }
 
+// 25.1.2.14 ArrayBufferCopyAndDetach ( arrayBuffer, newLength, preserveResizability ), https://tc39.es/proposal-arraybuffer-transfer/#sec-arraybuffer.prototype.transfertofixedlength
+ThrowCompletionOr<ArrayBuffer*> copy_and_detach(VM& vm, ArrayBuffer& array_buffer, Value new_length, PreserveResizability)
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp:186]: `array_buffer_copy_and_detach`

File: Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp [hunk 7]
```diff
@@ -182,4 +182,52 @@ ThrowCompletionOr<ArrayBuffer*> clone_array_buffer(VM& vm, ArrayBuffer& source_b
...
+{
+    auto& realm = *vm.current_realm();
+
+    // 1. Perform ? RequireInternalSlot(arrayBuffer, [[ArrayBufferData]]).
+
+    // FIXME: 2. If IsSharedArrayBuffer(arrayBuffer) is true, throw a TypeError exception.
+
+    // NOTE: This seems more logical to be before step 3, we should probably file a spec bug
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp:194]: No, the spec generally favours doing observable coercion of input arguments first (ToIndex in this case). Apply the suggestion from the first commit and the issue of accessing byte length goes away :)

**Review by linusg — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibJS/Print.cpp (hunk 1 above)
Comment (PrestonLTaylor) [reply Userland/Libraries/LibJS/Print.cpp:395]: Done, just made a seperate commit to prevent the crash when accessing byte_length from a detached array

**Review by linusg — APPROVED** [review]
LGTM!

**Pull request merged**